### ISSUE TITLE
sndfile: unknown format or could not open file - /w2l/data/train-clean-100/000002379.flac18

### ISSUE THREAD

**Issue by zdgithub** [facebookresearch/wav2letter#241]
When I try to train the librispeech, the following errors occurred:
```
F0319 22:54:32.525249 15292 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000000000.flac18
F0319 22:54:32.525315 15300 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000004758.flac0
F0319 22:54:32.525328 15308 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000023783.flac18
F0319 22:54:32.525352 15301 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000007137.flac18
F0319 22:54:32.525391 15302 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000009515.flac18
F0319 22:54:32.525421 15305 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000016649.flac18
F0319 22:54:32.525449 15299 Sound.cpp:62] sndfile: unknown format or could not open file - /home/zd/w2l/data/train-clean-100/000002379.flac18
...
...
*** Check failure stack trace: ***
...
    @     0x7f85844abe21  google::LogMessage::SendToLog()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a9c6d  google::LogMessage::Fail()
    @     0x7f85844abe21  google::LogMessage::SendToLog()
    @     0x7f85844abe21  google::LogMessage::SendToLog()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x555f7e3abfe4  _ZN3w2l23W2lNumberedFilesDataset15loadSampleSizesEv._omp_fn.0
    @     0x555f7e3abfe4  _ZN3w2l23W2lNumberedFilesDataset15loadSampleSizesEv._omp_fn.0
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844abe21  google::LogMessage::SendToLog()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f8583b54ecf  GOMP_parallel
    @     0x7f8583b5d97e  (unknown)
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x555f7e3ac3f1  w2l::W2lNumberedFilesDataset::loadSampleSizes()
    @     0x555f7e3abfe4  _ZN3w2l23W2lNumberedFilesDataset15loadSampleSizesEv._omp_fn.0
    @     0x555f7e3ad560  w2l::W2lNumberedFilesDataset::W2lNumberedFilesDataset()
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x555f7e38b894  w2l::createDataset()
    @     0x555f7e3abfe4  _ZN3w2l23W2lNumberedFilesDataset15loadSampleSizesEv._omp_fn.0
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85cdf6e6db  start_thread
    @     0x555f7e219ca8  main
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844a979d  google::LogMessage::Flush()
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f85844ac7d9  google::LogMessageFatal::~LogMessageFatal()
    @     0x555f7e3c19ba  speech::loadSoundInfo()
    @     0x7f858365f88f  clone
Aborted (core dumped)

```
I have installed the libsndfile.
And I know the number `18` or `0` at the end is from `sf_error(file)` in Sound.cpp:62
OS: 18.04
cuda: 10.1
gcc: 7.3.0

**Comment by manu51188** [comment]
@zdgithub  

I am facing similar error on CPU backend while testing on Ubuntu 16.04 gcc version 5.4.0 while training.

I am not able to read even a single flac file (000000000.flac) which was prepared from python scripts. I have even changed the read & write permissions of flac files just to make sure permissions wont be a problem. 

mohit@mohit-VirtualBox:~$ ./wav2letter/build/Train train --flagsfile /home/mohit/wav2letter/tutorials/1-librispeech_clean/train.cfg 
F0318 06:34:27.840395  3678 Sound.cpp:62] sndfile: unknown format or could not open file - /home/mohit/w2l/data/train-clean-100/000000000.flac
*** Check failure stack trace: ***
    @           0x7096c8  google::LogMessage::Fail()
    @           0x70960c  google::LogMessage::SendToLog()
    @           0x708f48  google::LogMessage::Flush()
    @           0x70ca6f  google::LogMessageFatal::~LogMessageFatal()
    @           0x589348  speech::loadSoundInfo()
    @           0x578fc7  _ZN3w2l23W2lNumberedFilesDataset15loadSampleSizesEv._omp_fn.0
    @     0x7fcdd1417038  __kmp_api_GOMP_parallel
    @           0x57950f  w2l::W2lNumberedFilesDataset::loadSampleSizes()
    @           0x57a698  w2l::W2lNumberedFilesDataset::W2lNumberedFilesDataset()
    @           0x564ad1  w2l::createDataset()
    @           0x41a48f  main
    @     0x7fcdd050e830  __libc_start_main
    @           0x463159  _start
    @              (nil)  (unknown)
Aborted (core dumped)

Can anyone help me resolve this issue ?

Regards,
Mohit

**Comment by jacobkahn** [comment]
@zdgithub — is libsndfile able to open other files from the dataset/does it only fail on these files?

**Comment by manu51188** [comment]
Hi Jacob kahn,

It looks like I am having the above mentioned problem in all the flac files which are read.

To verify this, I have tested the Librispeech clean files downloaded from the link "wget -qO- http://www.openslr.org/resources/12/dev-clean.tar.gz | tar xvz -C $W2LDIR" and and same error is coming with my test program as well.

sndfile: unknown format or could not open file - /home/mohit/1272-128104-0000.flac

Regards,
Mohit

**Comment by zdgithub** [comment]
> @zdgithub — is libsndfile able to open other files from the dataset/does it only fail on these files?

I  test some `.flac` files, but the error remains the same. 
Then I try testing a `.wav` file, it could read the format.
That is to say, the sndfile just couldn't read the `.flac` files.  I'm confused.

**Comment by jacobkahn** [comment]
@zdgithub @manu51188  — this sounds like it may be an issue with libsndfile. Can you try running libsndfile alone and opening some `flac` files? It's possible it's a recent bug with the library depending on when you built it. I'd consider using [an old release](https://github.com/erikd/libsndfile/tree/1.0.28) or building from source, then testing again.

**Comment by zdgithub** [comment]
> @zdgithub @manu51188 — this sounds like it may be an issue with libsndfile. Can you try running libsndfile alone and opening some `flac` files? It's possible it's a recent bug with the library depending on when you built it. I'd consider using [an old release](https://github.com/erikd/libsndfile/tree/1.0.28) or building from source, then testing again.

Now I try using [the old release](https://github.com/erikd/libsndfile/tree/1.0.28) libsndfile-1.0.28, but the error still exists.  And when `make check` the libsndfile, I got the following:
```
 libsndfile-1.0.28 passed tests on FLAC files.
----------------------------------------------------------------------
    ogg_short_test                 : vorbis_short.oga ........ ok
    ogg_seek_test                  : vorbis_short.oga ........ ok
    ogg_int_test                   : vorbis_int.oga .......... ok
    ogg_seek_test                  : vorbis_int.oga .......... ok
    ogg_float_test                 : vorbis_float.oga ........ ok
    ogg_seek_test                  : vorbis_float.oga ........ ok
    ogg_double_test                : vorbis_double.oga ....... ok
    ogg_seek_test                  : vorbis_double.oga ....... ok
    ogg_stereo_seek_test           : vorbis_seek.ogg ......... ok
    vorbis_test                    : vorbis_test.oga ......... ok
    compression_size_test          : q[36]_vorbis.oga ........ ok
    sdlcomp_test_short             : vorbis.oga .............. ok
    sdlcomp_test_int               : vorbis.oga .............. ok
    sdlcomp_test_float             : vorbis.oga .............. Not working for this format.
    sdlcomp_test_double            : vorbis.oga .............. Not working for this format.
    string_start_test              : vorbis.oga .............. ok
    zero_data_test                 : zerolen.oga ............. ok
----------------------------------------------------------------------

```
Is it the problem?

**Comment by manu51188** [comment]
Hi Jacob,

I am still not sure with libsndfile version as well as pointed out by @zdgithub.  

Can you please tell us the libsndfile version which we can use for reading of flac files ? This will help us move ahead because we are struck at this point.

Regards,
Mohit

**Comment by manu51188** [comment]
Hi @zdgithub  & @jacobkahn ,

I am able to resolve the .flac files reading error by building the libsnd library with libopus package.

Here are the details to how to build libopus package:
"https://github.com/xiph/opus".

Regards,
Mohit

**Comment by zdgithub** [comment]
@manu51188  ok, I will have a try.
Thank you all!

**Comment by kremnik** [comment]
Hi @manu51188 @zdgithub!
Could you please tell how did you build libsnd with libopus and did it solve the problem?

Best Regards,
Alexander

**Comment by manu51188** [comment]
Hi Alexander,

Sorry for my late reply. Atleast I am able to solve that error.

We require to build libsndfile with Ogg,Vorbis, flac file enables but to build libsndfile with all mentioned flags we require libopus package as well. If the libopus is not available or if version of libopus package < 1.2 then also these flags will not enabled. This can also be checked when you run ./configure command in which warning will come if libopus package is missing or is not of required version.

I have built the libopus package from source whose github link is given above.

Let me know if you still face any problem.
